README.md
STATUS
======

Stud is now officially abandonware, thanks for playing.

Recommended alternative: https://github.com/varnish/hitch

More info: https://blog.jamwt.com/2016/03/21/stud-no-more/

stud - The Scalable TLS Unwrapping Daemon
=========================================

`stud` is a network proxy that terminates TLS/SSL connections and forwards the
unencrypted traffic to some backend.  It's designed to handle 10s of thousands of
connections efficiently on multicore machines.

It follows a process-per-core model; a parent process spawns N children who
each `accept()` on a common socket to distribute connected clients among them.
Within each child, asynchronous socket I/O is conducted across the local
connections using `libev` and `OpenSSL`'s nonblocking API.  By default,
`stud` has an overhead of ~200KB per connection--it preallocates
some buffer space for data in flight between frontend and backend.

`stud` has very few features--it's designed to be paired with an intelligent
backend like haproxy or nginx.  It maintains a strict 1:1 connection pattern
with this backend handler so that the backend can dictate throttling behavior,
maxmium connection behavior, availability of service, etc.

`stud` will optionally write the client IP address as the first few octets
(depending on IPv4 or IPv6) to the backend--or provide that information
using HAProxy's PROXY protocol.  When used with the PROXY protocol, `stud` can
also transparently pass an existing PROXY header to the cleartext stream.  This
is especially useful if a TCP proxy is used in front of `stud`.  Using either of
these techniques, backends who care about the client IP can still access it even
though `stud` itself appears to be the connected client.

Thanks to a contribution from Emeric at Exceliance (the folks behind HAProxy),
a special build of `stud` can be made that utilitizes shared memory to
use a common session cache between all child processes.  This can speed up
large `stud` deployments by avoiding client renegotiation.

Releases
---------

Please be aware of the policy regarding releases, code stability, and security:

 * In git, the tip of the master branch should always build on Linux and
   FreeBSD, and is likely to be as stable as any other changeset.  A
   careful review of patches is conducted before being pushed to github.
 * Periodically, a version tag will be pushed to github for an old(er)
   changeset--0.1, 0.2, etc.  These tags mark a particular release of
   `stud` that has seen heavy testing and several weeks of production
   stability.  Conservative users are advised to use a tag.
 * `stud` has an optional build that utilizes shared memory-based SSL contexts
   and UDP peer communication to keep a session cache between many child processes
   running on many machines.  The use of this build can dramatically speed
   up SSL handshakes on many-core and/or clustered deployments.
   However, it's important to acknowledge the inevitable theoretical
   security tradeoffs associated with the use of this (substantially more
   complex) binary.  Therefore, the deeply paranoid are advised to use
   only the standard `stud` binary at the cost of some performance.

Requirements and Limitations
----------------------------

`stud` requires:

    libev >= 4
    openssl (recent, >=1.0.0 recommended)

Stud currently works on Linux, OpenBSD, FreeBSD, and MacOSX.
It has been tested the most heavily on Linux/x86_64.

While porting it to other POSIX platforms is likely trivial, it hasn't be done
yet. Patches welcome!

If you're handling a large number of connections, you'll
probably want to raise `ulimit -n` before running `stud`.
It's very strongly recommended to not run `stud` as root; ideally, it would
be run as a user ("stud", perhaps) that does nothing but run `stud`.  Stud
will setuid (using -u) after binding if you need to bind to a low port (< 1024).

Installing
----------

To install `stud`:

    $ make
    $ sudo make install

Usage
-----

The only required argument is a path to a PEM file that contains the certificate
(or a chain of certificates) and private key. If multiple certificates are
given, `stud` will attempt to perform SNI (Server Name Indication) on new
connections, by comparing the indicated name with the names on each of the
certificates, in order. The first certificate that matches will be used. If none
of the certificates matches, the last certificate will be used as the default.

Detail about the entire set of options can be found by invoking `stud -h`:

    CONFIGURATION:

            --config=FILE      Load configuration from specified file.
            --default-config   Prints default configuration to stdout.

    ENCRYPTION METHODS:

          --tls                   TLSv1 (default)
          --ssl                   SSLv3 (implies no TLSv1)
      -c  --ciphers=SUITE         Sets allowed ciphers (Default: "")
      -e  --ssl-engine=NAME       Sets OpenSSL engine (Default: "")
      -O  --prefer-server-ciphers Prefer server list order

    SOCKET:

      -b  --backend=HOST,PORT     Backend [connect] (default is "[127.0.0.1]:8000")
      -f  --frontend=HOST,PORT    Frontend [bind] (default is "[*]:8443")

    PERFORMANCE:

      -n  --workers=NUM          Number of worker processes (Default: 1)
      -B  --backlog=NUM          Set listen backlog size (Default: 100)
          --max-open-files=NUM   Set maximum open files before (Default: 1024)
      -k  --keepalive=SECS       TCP keepalive on client socket (Default: 3600)

    SECURITY:

      -r  --chroot=DIR           Sets chroot directory (Default: "")
      -u  --user=USER            Set uid/gid after binding the socket (Default: "")
      -g  --group=GROUP          Set gid after binding the socket (Default: "")

    LOGGING:
      -q  --quiet                Be quiet; emit only error messages
      -s  --syslog               Send log message to syslog in addition to stderr/stdout
      --syslog-facility=FACILITY Syslog facility to use (Default: "daemon")

    OTHER OPTIONS:
          --daemon               Fork into background and become a daemon (Default: off)
          --write-ip             Write 1 octet with the IP family followed by the IP
                                 address in 4 (IPv4) or 16 (IPv6) octets little-endian
                                 to backend before the actual data
                                 (Default: off)
          --write-proxy          Write HaProxy's PROXY (IPv4 or IPv6) protocol line
                                 before actual data
                                 (Default: off)
          --proxy-proxy          Proxy HaProxy's PROXY (IPv4 or IPv6) protocol line
                                 before actual data
                                 (Default: off)

      -t  --test                 Test configuration and exit
      -V  --version              Print program version and exit
      -h  --help                 This help message

Configuration File
------------------

Stud can also use a configuration file that supports all the same options as the
command-line arguments. You can use `stud --default-config` to
generate the default configuration on stdout; then, customize your configuration and
pass it to `stud --config=FILE`.

Serving HTTPS
-------------

If you're using `stud` for HTTPS, please make sure to use the `--ssl` option!


Diffie–Hellman
--------------

To use DH with stud, you will need to add some bytes to your pem file:

% openssl dhparam -rand - 1024 >> PEMFILE

Be sure to set your cipher suite appropriately: -c DHE-RSA-AES256-SHA

Authors
-------

`stud` was originally written by Jamie Turner (@jamwt) and is maintained
by the Bump (http://bu.mp) server team.  It currently (12/11) provides
server-side TLS termination for over 85 million Bump users.

Special thanks to Colin Percival (@cperciva) for an early security
audit and code review.

Finally, thank you to all the stud contributors, who have taken the
program from a good start to a solid project:

https://github.com/bumptech/stud/contributors


debian/changelog
stud (0.3~0.2014111901~wheezy) unstable; urgency=low

  * Doc -y and --openssl-cache-size in usage

 -- Stefan Wehner <stefan@tuenti.com>  Wed, 19 Nov 2014 10:27:48 +0000

stud (0.3~0.2014101501~wheezy) unstable; urgency=low

  * Added the option to disable SSLv3.

 -- Alfredo Beaumont <abeumont@tuenti.com>  Wed, 15 Oct 2014 10:39:49 +0000

stud (0.3~0.2014091601~wheezy) unstable; urgency=low

  * Added -x and --session-timeout parameters in order to control OpenSSL TLS
    session timeout which is 300 seconds by default.

 -- Ricardo Bartolome Mendez <ricardo@tuenti.com>  Tue, 16 Sep 2014 09:45:15 +0000

stud (0.3~0.2014062401~wheezy) unstable; urgency=low

  * Merged ecdh_named_curves from https://github.com/b/stud

 -- Ricardo Bartolome Mendez <ricardo@tuenti.com>  Tue, 24 Jun 2014 12:44:13 +0000

stud (0.3~0.2014042801~wheezy) unstable; urgency=low

  * Build for Debian Wheezy

 -- Ricardo Bartolome Mendez <ricardo@tuenti.com>  Mon, 28 Apr 2014 12:50:35 +0000

stud (0.3~0.2014042801~squeeze) unstable; urgency=low

  * Rev 150c3545458f1a40939854890ee6d5eb31a1383b linked against ebtree 6.0.8
  * Built for Debian Squeeze

 -- Ricardo Bartolome Mendez <ricardo@tuenti.com>  Mon, 28 Apr 2014 12:44:52 +0000

stud (0.3~0.20140428) unstable; urgency=low

  * Non-maintainer upload.
  * Built revision 150c3545458f1a40939854890ee6d5eb31a1383b

 -- Ricardo Bartolome Mendez <ricardo@tuenti.com>  Mon, 28 Apr 2014 12:32:20 +0000

stud (0.3~0.20111212) unstable; urgency=low

  * Initial release.

 -- Charlie Root <root@interseek.com>  Mon, 12 Dec 2011 13:24:18 +0100



log.h
/**
 * log.h
 *
 * Author: Brane F. Gracnar
 *
 */

// "Initializes" logging subsystem
void log_init (void);

// Writes timestamp (including microseconds) to dst.
// If dst == NULL uses it's own buffer. 
char * ts_as_str (char *dst);

// Writes string representation of proxystate structure to dst.
// If dst == NULL uses it's own buffer.
char * ps_as_str (proxystate *ps, char *dst);

// Writes string representation of a socket to dst.
// If dst == NULL uses it's own buffer.
char * sock_as_str (int sock, char *dst);

char * log_level2str (int level);
int log_str2level (char *str);

/**
 * Log level names
 */

#define LSTR_FATAL      "FATAL"
#define LSTR_ERR        "ERROR"
#define LSTR_WARN       "WARNING"
#define LSTR_NOTICE     "NOTICE"
#define LSTR_INFO       "INFO"
#define LSTR_DEBUG      "DEBUG"
#define LSTR_UNKNOWN    "UNKNOWN"

/**
 * General purpose logging functions
 */

void log_msg (int level, char *fmt, ...);

#define log_fatal(...)          log_msg(LOG_EMERG, __VA_ARGS__)
#define log_err(...)            log_msg(LOG_ERR, __VA_ARGS__)
#define log_warn(...)           log_msg(LOG_WARNING, __VA_ARGS__)
#define log_notice(...)         log_msg(LOG_NOTICE, __VA_ARGS__)
#define log_info(...)           log_msg(LOG_INFO, __VA_ARGS__)
#define log_debug(...)          log_msg(LOG_DEBUG, __VA_ARGS__)

/**
 * Proxy state logging "functions"
 */

void log_msg_ps (int level, proxystate *ps, char *fmt, ...);

#define log_fatal_ps(...)        log_msg_ps(LOG_EMERG, __VA_ARGS__)
#define log_err_ps(...)          log_msg_ps(LOG_ERR, __VA_ARGS__)
#define log_warn_ps(...)         log_msg_ps(LOG_WARNING, __VA_ARGS__)
#define log_notice_ps(...)       log_msg_ps(LOG_NOTICE, __VA_ARGS__)
#define log_info_ps(...)         log_msg_ps(LOG_INFO, __VA_ARGS__)
#define log_debug_ps(...)        log_msg_ps(LOG_DEBUG, __VA_ARGS__)

/**
 * Socket logging "functions"
 */

void log_msg_sock (int level, int sock, char *fmt, ...);

#define log_fatal_sock(...)      log_msg_sock(LOG_EMERG, __VA_ARGS__)
#define log_err_sock(...)        log_msg_sock(LOG_ERR, __VA_ARGS__)
#define log_warn_sock(...)       log_msg_sock(LOG_WARNING, __VA_ARGS__)
#define log_notice_sock(...)     log_msg_sock(LOG_NOTICE, __VA_ARGS__)
#define log_info_sock(...)       log_msg_sock(LOG_INFO, __VA_ARGS__)
#define log_debug_sock(...)      log_msg_sock(LOG_DEBUG, __VA_ARGS__)

/**
 * Socket and proxy state logging functions
 */

void log_msg_lps (int level, int listener, proxystate *ps, char *fmt, ...);

#define log_fatal_lps(...)      log_msg_lps(LOG_EMERG, __VA_ARGS__)
#define log_err_lps(...)        log_msg_lps(LOG_ERR, __VA_ARGS__)
#define log_warn_lps(...)       log_msg_lps(LOG_WARNING, __VA_ARGS__)
#define log_notice_lps(...)     log_msg_lps(LOG_NOTICE, __VA_ARGS__)
#define log_info_lps(...)       log_msg_lps(LOG_INFO, __VA_ARGS__)
#define log_debug_lps(...)      log_msg_lps(LOG_DEBUG, __VA_ARGS__)


/**
 * Fatal logging "functions"
 */

void fail (char *fmt, ...);

#define die(...)                                        \
  log_msg(LOG_CRIT, __VA_ARGS__);                      \
  exit(1)

#define die_ps(...)                                     \
  log_msg_ps(LOG_CRIT, __VA_ARGS__);                   \
  exit(1)

#define die_sock(...)                                   \
  log_msg_sock(LOG_CRIT, __VA_ARGS__);                 \
  exit(1)


stud.h
/**
  * Copyright 2011 Bump Technologies, Inc. All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without modification, are
  * permitted provided that the following conditions are met:
  *
  *    1. Redistributions of source code must retain the above copyright notice, this list of
  *       conditions and the following disclaimer.
  *
  *    2. Redistributions in binary form must reproduce the above copyright notice, this list
  *       of conditions and the following disclaimer in the documentation and/or other materials
  *       provided with the distribution.
  *
  * THIS SOFTWARE IS PROVIDED BY BUMP TECHNOLOGIES, INC. ``AS IS'' AND ANY EXPRESS OR IMPLIED
  * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
  * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL BUMP TECHNOLOGIES, INC. OR
  * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
  * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
  * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
  * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
  * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
  * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  *
  * The views and conclusions contained in the software and documentation are those of the
  * authors and should not be interpreted as representing official policies, either expressed
  * or implied, of Bump Technologies, Inc.
  *
  **/

#include <sys/socket.h>
#include <openssl/ssl.h>
#include <ev.h>

#include "ringbuffer.h"

/* What agent/state requests the shutdown--for proper half-closed
 * handling */
typedef enum _SHUTDOWN_REQUESTOR {
    SHUTDOWN_HARD,
    SHUTDOWN_DOWN,
    SHUTDOWN_UP
} SHUTDOWN_REQUESTOR;

/*
 * Proxied State
 *
 * All state associated with one proxied connection
 */
typedef struct proxystate {
    ringbuffer ring_down; /* pushing bytes from client to backend */
    ringbuffer ring_up;   /* pushing bytes from backend to client */

    ev_io ev_r_up;        /* Upstream write event */
    ev_io ev_w_up;        /* Upstream read event */

    ev_io ev_r_handshake; /* Downstream write event */
    ev_io ev_w_handshake; /* Downstream read event */

    ev_io ev_r_down;      /* Downstream write event */
    ev_io ev_w_down;      /* Downstream read event */

    int fd_up;            /* Upstream (client) socket */
    int fd_down;          /* Downstream (backend) socket */

    int want_shutdown:1;  /* Connection is half-shutdown */
    int handshaked:1;     /* Initial handshake happened */
    int renegotiation:1;  /* Renegotation is occuring */

    SSL *ssl;             /* OpenSSL SSL state */

    struct sockaddr_storage remote_ip;  /* Remote ip returned from `accept` */
} proxystate;

#define NULL_DEV "/dev/null"

char * stud_instance_name (char *name);


shctx.h
/*
 * shctx.h
 *
 * Copyright (C) 2011 EXCELIANCE
 *
 * Author: Emeric Brun - emeric@exceliance.fr
 *
 */

#ifndef SHCTX_H
#define SHCTX_H
#include <openssl/ssl.h>
#include <stdint.h>

#ifndef SHSESS_MAX_FOOTER_LEN
#define SHSESS_MAX_FOOTER_LEN sizeof(uint32_t) \
				+ EVP_MAX_MD_SIZE
#endif

#ifndef SHSESS_MAX_DATA_LEN
#define SHSESS_MAX_DATA_LEN 512
#endif

#define SHSESS_MAX_ENCODED_LEN SSL_MAX_SSL_SESSION_ID_LENGTH \
				+ SHSESS_MAX_DATA_LEN \
				+ SHSESS_MAX_FOOTER_LEN


/* Callback called on a new session event:
 * session contains the sessionid zeros padded to SSL_MAX_SSL_SESSION_ID_LENGTH
 *                                               followed by ASN1 session encoding.
 * len is set to SSL_MAX_SSL_SESSION_ID_LENGTH + ASN1 session length
 * len is always less than SSL_MAX_SSL_SESSION_ID_LENGTH + SHSESS_MAX_DATA_LEN.
 * Remaining Bytes from len to SHSESS_MAX_ENCODED_LEN can be used to add a footer.
 * cdate is the creation date timestamp.
 */
void shsess_set_new_cbk(void (*func)(unsigned char *session, unsigned int len, long cdate));

/* Add a session into the cache,
 * session contains the sessionid zeros padded to SSL_MAX_SSL_SESSION_ID_LENGTH
 *                                             followed by ASN1 session encoding.
 * len is set to SSL_MAX_SSL_SESSION_ID_LENGTH + ASN1 data length.
 *            if len greater than SHSESS_MAX_ENCODED_LEN, session is not added.
 * if cdate not 0, on get events session creation date will be reset to cdate */
void shctx_sess_add(const unsigned char *session, unsigned int session_len, long cdate);

/* Init shared memory context if not allocated and set SSL context callbacks
 * size is the max number of stored session
 * Returns: -1 on alloc failure, size if performs context alloc, and 0 if just
 * perform callbacks registration */
int shared_context_init(SSL_CTX *ctx, int size);

#endif /* SHCTX_H */


configuration.h
/**
 * configuration.h
 *
 * Author: Brane F. Gracnar
 *
 */

#include <sys/types.h>

#ifdef USE_SHARED_CACHE
  #include "shctx.h"

  #ifndef MAX_SHCUPD_PEERS
    #define MAX_SHCUPD_PEERS 15
  #endif

typedef struct shcupd_peer_opt {
     char *ip;
     char *port;
} shcupd_peer_opt;

#endif

typedef enum {
    ENC_TLS,
    ENC_SSL
} ENC_TYPE;

typedef enum {
    SSL_SERVER,
    SSL_CLIENT
} PROXY_MODE;

struct cert_files {
    char *CERT_FILE;
    struct cert_files *NEXT;
};

/* configuration structure */
typedef struct stud_config_s {
    ENC_TYPE ETYPE;
    PROXY_MODE PMODE;
    int WRITE_IP_OCTET;
    int WRITE_PROXY_LINE;
    int PROXY_PROXY_LINE;
    char *CHROOT;
    uid_t UID;
    gid_t GID;
    char *FRONT_IP;
    char *FRONT_PORT;
    char *BACK_IP;
    char *BACK_PORT;
    long NCORES;
    struct cert_files *CERT_FILES;
    char *CIPHER_SUITE;
    char *ENGINE;
    int BACKLOG;
#ifdef USE_SHARED_CACHE
    int SHARED_CACHE;
    char *SHCUPD_IP;
    char *SHCUPD_PORT;
    shcupd_peer_opt SHCUPD_PEERS[MAX_SHCUPD_PEERS+1];
    char *SHCUPD_MCASTIF;
    char *SHCUPD_MCASTTTL;
#endif
    int QUIET;
    int SYSLOG;
    int SYSLOG_FACILITY;
    int TCP_KEEPALIVE_TIME;
    int DAEMONIZE;
    int PREFER_SERVER_CIPHERS;
} stud_config;

char * config_error_get (void);
stud_config * config_new (void);
void config_destroy (stud_config *cfg);
int config_file_parse (char *file, stud_config *cfg);
void config_parse_cli(int argc, char **argv, stud_config *cfg);


log.c
/**
 * log.c
 *
 * Author: Brane F. Gracnar
 *
 */

#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>

#include <ev.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>

#include <openssl/x509.h>
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/engine.h>
#include <errno.h>
#include <syslog.h>


#include "ringbuffer.h"
#include "stud.h"
#include "configuration.h"
#include "log.h"


#define TIMESTR_BUF_LEN         40
#define TIMESTR_FMT             "%Y/%m/%d %H:%M:%S.%%d"
#define SOCKSTR_BUF_LEN         48
#define PSSTR_BUF_LEN           100
#define LOG_BUF_LEN             1024

#define DOLOG(l) (l <= CONFIG->LOG_LEVEL) ? 1 : 0


char buf_timestr[TIMESTR_BUF_LEN];
char sockstr_buf[SOCKSTR_BUF_LEN];
char psstr_buf[PSSTR_BUF_LEN];

static int logging_initialized = 0;

extern stud_config *CONFIG;
extern int child_num;

char * ts_as_str (char *dst) {
  if (dst == NULL) dst = buf_timestr;

  char fmtbuf[TIMESTR_BUF_LEN];
  memset(fmtbuf, '\0', sizeof(fmtbuf));
  memset(dst, '\0', TIMESTR_BUF_LEN);

  // get cached loop time...
  ev_tstamp t_ev = ev_now(EV_DEFAULT);

  time_t t = (time_t) t_ev;
  struct tm *lt = localtime(&t);
  strftime(fmtbuf, sizeof(fmtbuf), TIMESTR_FMT, lt);

  // add microtime
  int frac = (t_ev - (long) t_ev) * 1000000;
  snprintf(dst, TIMESTR_BUF_LEN, fmtbuf, frac);

  return dst;
}

char * sock_as_str (int sock, char *dst) {
  if (dst == NULL) dst = sockstr_buf;
  memset(dst, '\0', SOCKSTR_BUF_LEN);

  struct sockaddr_storage addr;
  char ipstr[INET6_ADDRSTRLEN];
  int port = 0;

  socklen_t len = sizeof(addr);
  getpeername(sock, (struct sockaddr*)&addr, &len);
  
  // IPv4 socket?
  if (addr.ss_family == AF_INET) {
    struct sockaddr_in *s = (struct sockaddr_in *)&addr;
    port = ntohs(s->sin_port);
    inet_ntop(AF_INET, &s->sin_addr, ipstr, sizeof(ipstr));
  }
  // IPv6 socket?
  else if (addr.ss_family == AF_INET6) {
    struct sockaddr_in6 *s = (struct sockaddr_in6 *)&addr;
    port = ntohs(s->sin6_port);
    inet_ntop(AF_INET6, &s->sin6_addr, ipstr, sizeof(ipstr));
  }
  // UNIX domain socket?
  else if (addr.ss_family == AF_UNIX) {
    
  }

  snprintf(dst, SOCKSTR_BUF_LEN, "[%s]:%d", ipstr, port);
  return dst;
}

char * ps_as_str (proxystate *ps, char *dst) {
  char up_buf[SOCKSTR_BUF_LEN];
  char down_buf[SOCKSTR_BUF_LEN];

  if (dst == NULL) dst = psstr_buf;
  memset(dst, '\0', PSSTR_BUF_LEN);
    
  sock_as_str(ps->fd_up, up_buf);
  sock_as_str(ps->fd_down, down_buf);
  
  snprintf(dst, PSSTR_BUF_LEN, "%s => %s", up_buf, down_buf);
  
  return dst;
}

char buf_pcsig[10];
char * proc_signature (char *dst) {
  if (child_num < 0 || child_num >= CONFIG->NCORES)
    return "master";
  else {
    if (dst == NULL) dst = buf_pcsig;
    
    // already cached stuff?
    if (dst == buf_pcsig && buf_pcsig[0] != '\0') return dst;
    
    memset(dst, '\0', sizeof(buf_pcsig));
    snprintf(dst, sizeof(buf_pcsig), "worker #%d", child_num);
  }
  
  return dst;
}

char * log_level2str (int level) {
  switch (level) {
    case LOG_EMERG:
      return LSTR_FATAL;
      break;
    case LOG_ERR:
      return LSTR_ERR;
      break;
    case LOG_WARNING:
      return LSTR_WARN;
      break;
    case LOG_NOTICE:
      return LSTR_NOTICE;
      break;
    case LOG_INFO:
      return LSTR_INFO;
      break;
    case LOG_DEBUG:
      return LSTR_DEBUG;
      break;
    default:
      return LSTR_UNKNOWN;
  }
}

int log_str2level (char *str) {
  int r = LOG_NOTICE;
  if (str == NULL || strlen(str) < 1) return r;
  
  if (strcasecmp(str, LSTR_FATAL) == 0)
    r = LOG_EMERG;
  else if (strcasecmp(str, LSTR_ERR) == 0)
    r = LOG_ERR;
  else if (strcasecmp(str, LSTR_WARN) == 0 || strcasecmp(str, "warn") == 0)
    r = LOG_WARNING;
  else if (strcasecmp(str, LSTR_NOTICE) == 0)
    r = LOG_NOTICE;
  else if (strcasecmp(str, LSTR_INFO) == 0)
    r = LOG_INFO;
  else if (strcasecmp(str, LSTR_DEBUG) == 0)
    r = LOG_DEBUG;

  return r;
}

char buf_syslog_ident[128];
char * log_syslog_ident (void) {
  char *n = stud_instance_name(NULL);
  if (n == NULL || strlen(n) < 1)
    return "stud";
  
  snprintf(buf_syslog_ident, sizeof(buf_syslog_ident), "stud/%s", n);
  return buf_syslog_ident;
}

void log_init (void) {
  if (logging_initialized) return;
  if (CONFIG->SYSLOG) {
    openlog(log_syslog_ident(), LOG_CONS | LOG_PID | LOG_NDELAY, CONFIG->SYSLOG_FACILITY);
  }
}

void log_msg (int level, char *fmt, ...) {
  va_list args;
  char buf[LOG_BUF_LEN];
  
  //printf("LCFG LLEVEL: %d, LLEVEL: %d\n", CONFIG->LOG_LEVEL, level);
  if (! DOLOG(level)) return;
  //printf("DAEMONIZE: %d\n", CONFIG->DAEMONIZE);
  //printf("SYSLOG: %d\n", CONFIG->SYSLOG);
  if (CONFIG->DAEMONIZE && ! CONFIG->SYSLOG) return;

  // format log message
  memset(buf, '\0', LOG_BUF_LEN);
  va_start(args, fmt);
  vsnprintf(buf, sizeof(buf), fmt, args);
  va_end(args);

  if (! CONFIG->DAEMONIZE) {
    fprintf(stderr, "[%s] %-10s %-7s %s\n", ts_as_str(NULL), proc_signature(NULL), log_level2str(level), buf);
  }
  
  if (CONFIG->SYSLOG) {
    syslog(level, "%s %s", proc_signature(NULL), buf);
  }
}


void log_msg_ps (int level, proxystate *ps, char *fmt, ...) {
  va_list args;
  char buf[LOG_BUF_LEN];
  
  if (! DOLOG(level)) return;

  // format log message
  memset(buf, '\0', LOG_BUF_LEN);
  va_start(args, fmt);
  vsnprintf(buf, sizeof(buf), fmt, args);
  va_end(args);
  
  log_msg(level, "%s %s", ps_as_str(ps, NULL), buf);
}

void log_msg_sock (int level, int sock, char *fmt, ...) {
  va_list args;
  char buf[LOG_BUF_LEN];
  
  if (! DOLOG(level)) return;

  // format log message
  memset(buf, '\0', LOG_BUF_LEN);
  va_start(args, fmt);
  vsnprintf(buf, sizeof(buf), fmt, args);
  va_end(args);
  
  log_msg(level, "%s %s", sock_as_str(sock, NULL), buf);
}

void log_msg_lps (int level, int listener, proxystate *ps, char *fmt, ...) {
  va_list args;
  char buf[LOG_BUF_LEN];
  
  if (! DOLOG(level)) return;

  // format log message
  memset(buf, '\0', LOG_BUF_LEN);
  va_start(args, fmt);
  vsnprintf(buf, sizeof(buf), fmt, args);
  va_end(args);
  
  log_msg(
    level, "[%s %s] %s", sock_as_str(listener, NULL),
    ps_as_str(ps, NULL), buf
  );  
}

void fail (char *fmt, ...) {
  va_list args;
  char buf[LOG_BUF_LEN];

  // format log message
  memset(buf, '\0', LOG_BUF_LEN);
  va_start(args, fmt);
  vsnprintf(buf, sizeof(buf), fmt, args);
  va_end(args);
  
  die("%s %s", buf, strerror(errno));
}
